### PR TITLE
Escape discouraged control codes in text and attributes

### LINKED ISSUE #123: Escape 'highly discouraged' control characters
When generating XML, depending on the version, some characters are forbidden or otherwise `highly discouraged`.

See https://www.w3.org/TR/xml/#charsets

I suggest to at least escape all the single byte characters from the [C0 and C1 control codes](https://en.wikipedia.org/wiki/C0_and_C1_control_codes) except the explicitly listed NUL (0), HT (tab), LF (linefeed), CR (carriage return) and NEL (next line).

Possible also escape the listed high unicode characters.

### PR DESCRIPTION
**Pull request by rspilker**
Fixes #123

### CODE REVIEW

**Comment by renggli** [comment]
Should this go into a new `XmlEntityMapping` class? This seems to be quite a big change I am unsure about its impact on other users? I definitely never saw the need for this myself. Could you add a test exercising the behavior in a broader context (parsing/writing and creating/writing an XML file)?

**Comment by rspilker** [comment]
I've thought about a different XmlEntityMapping class, but the specification is quite clear: you should not do that.

Also, any XML-consuming code should be fine with accepting &#x3; instead of a `\u0003`. 

I did encounter a problem with live code where it was the other way around: the consumer could not handle `\u0003`.

Regarding the tests, I did do some testing on my own copy. I thought about parsing an existing file, and than parse the `toXmlString()`-ed version. They should always be the same. Do you have code that can see if two documents are equal? I mean without calling `toXmlString()`?

**Comment by rspilker** [comment]
I just thought: all existing tests already give us quite some indication that for xml files that do not contain any control characters, a still goes well. I did not need to change any of the existing tests after modifying the behavior.

So I could insert a control character, for example `\0007`, into the `booksXml` in `test/examples.dart` and then let the invariant check take care of most of the heavy lifting

**Comment by renggli** [comment]
Ok, I agree that there is no need for a new entity mapping. I was actually quite surprised that existing tests did not fail 😊

Maybe you could add [a new example](https://github.com/renggli/dart-xml/blob/main/test/examples.dart#L213) (like the unicode one) and [run the invariants](https://github.com/renggli/dart-xml/blob/main/test/examples_test.dart) over it?

File: test/entity_test.dart [hunk 1]
```diff
@@ -53,6 +53,8 @@ void testDefaultMapping(XmlEntityMapping entityMapping) {
     test('text', () {
       expect(entityMapping.encodeText('<'), '&lt;');
       expect(entityMapping.encodeText('&'), '&amp;');
+      expect(entityMapping.encodeText('\0\u0008\u0009\u0084\u0085\u0086'),
```
[the remark this answers is not in the document]
Comment (rspilker) [reply test/entity_test.dart]: Yes I saw it and is committed https://github.com/renggli/dart-xml/pull/124/commits/17c2765df857f6e2ae3ca379e298090d634c1f64 to fix it

File: lib/src/xml/entities/default_mapping.dart [hunk 2]
```diff
@@ -57,60 +57,63 @@ class XmlDefaultEntityMapping extends XmlEntityMapping {
...
-      return '&#xD;';
-    case '\t':
-      return '&#x9;';
+    default:
+      return _asNumericCharacterReferences(toEscape!);
   }
-  throw ArgumentError.value(match, 'match');
 }
+
+// Lists all C0 and C1 control codes except NUL, HT, LF, CR and NEL
```
[the remark this answers is not in the document]
Comment (rspilker) [reply lib/src/xml/entities/default_mapping.dart:113]: My source was https://www.w3.org/TR/xml11/#charsets and then `[2a]  RestrictedChar`.

But now I also found https://www.w3.org/TR/xml/#charsets which is even stronger

[the remark this answers is not in the document]
Comment (rspilker) [reply lib/src/xml/entities/default_mapping.dart:113]: https://en.wikipedia.org/wiki/XML#Valid_Unicode_characters_in_XML_1.0_and_XML_1.1

**Comment by renggli** [comment]
Thank you!

**Pull request merged**